### PR TITLE
Add experimental GDK GUI backend

### LINKED ISSUE #1729: System keyboard layout is ignored on Windows
FontForge always uses QWERTY apparently.

cc @jtanx

### LINKED ISSUE #2096: Mniature pop-up windows
miniature pop-up windows on taskbar show the same image.
i expect to see a thumbnail of each opened window of FF like in other applications. 

when taken the curser over them.
on windows 7 latest stable fontforge 

img_20150302_142914

### LINKED ISSUE #2256: Windows open off-screen on multi-monitor setup (Windows)
When opening a font, that window is moved to what FontForge believes is the upper-right corner of the screen. Apparently, it does this by looking at the total dimensions of all connected monitors, not the dimensions of any single monitor. Consequently, the window may be positioned in an area that is not actually visible. 

Example/to recreate: I run a three-monitor setup with a large QFHD monitor in the center and a smaller FHD monitor at the right, bottom aligned with the bottom of the center monitor. The font window is opened in the non-imaged space above the right-hand FHD monitor.

Workaround: while FontForge's windows do not respond to the usual Alt+Space, m shortcut to move windows, they do respond to Win+left arrow and Win+right arrow keyboard shortcuts will move them to visible areas.

OS: Windows 8.1 Pro x64
Version: 09:31 AWST 1-May-2015

### LINKED ISSUE #2461: Pulldown menus can become detached from main menu (Windows)
When opening a pull-down menu (for example the "File" menu) and then dragging the program window will detach the pull-down menu and leave it floating on the screen:
menu_detach
Expected behavior would be that the pull-down menu is automatically being closed if the program window is being dragged.

Related to this issue is the undesired behavior of FontForge of not closing an open pull-down menu when another application window is being activated (by mouse-click or Alt+TAB).

OS: MS Windows 7 Pro x64
FontForge version: FontForge-mingw-w64-i686-99da6f-r2

### LINKED ISSUE #2462: Menus shown behind taskbar (Windows)
If the FontForge window is placed at an unfortunate screen location, menus will open behind the taskbar, thus making the bottom menu entry/entries inaccessible until the FontForge window is being moved and the menu reopened:
taskbar
The pull-down menus should either always being displayed on top of the taskbar, or alternatively the space occupied by the taskbar needs to be taken into account when deciding whether pull-down menus should open downwards or upwards.

Related to issue #2256:
When deciding whether pull-down menus should open upwards or downwards, FontForge bases this decision incorrectly on the dimension of the virtual-screen bounding-box. This can lead to situations where a pull-down menu is being drawn off-screen:
menu_cut
The situation as displayed in this screenshot can for example occur with the following multi-monitor setup when FontForge is placed close to the bottom of monitor 1:
mm

OS: MS Windows 7 Pro x64
FontForge version: FontForge-mingw-w64-i686-99da6f-r2

### LINKED ISSUE #2590: How to change window's title name from VcXsrv X ??
I want to change window's title name from VcXsrv, but I couldn't find window title function, where is it.
Could you answer Where is it??

### PR DESCRIPTION
**Pull request by jtanx**
Note: Not to be merged as of yet.

---

This adds an experimental GDK backend to GDraw. It is meant to be more or less a drop-in replacement for the X11 backend. This backend is enabled by passing `--enable-gdk` at the configure stage. This will add a dependency on `libgdk-3.0` or `libgdk-2.0` (provided by `libgtk-3-dev` and `libgtk2.0-dev` or equivalent respectively) - it will prefer `libgdk-3.0` to `libgdk-2.0`, but either should work. If you wish, you may force it to explicitly build with either GDK2 or GDK3 by passing `--enable-gdk=gdk2` or `--enable-gdk=gdk3`.
### Motivation and Context

By using the GDK backend, it removes the hard dependency on X11. This is beneficial on Windows and OS X as it removes the need to run an X server (VcXsrv or XQuartz). Note that most of the benefits of this backend would be on Windows.

Side-effects of using GDK:
- Better multi-monitor support - windows should no longer open halfway between monitors  - fixes #2256 and #2381
- Menus open in front of the taskbar - fixes #2462. Also fixes #2461 as a side-effect of not being able to move the window without first dismissing the menu.
- Proper centering of windows on screen (by taking into account the size of window decorations) and not opening windows under the cursor (not really a side-effect since I specifically coded this in)
- Native support for the user's current keyboard, including IME support (IME support on Windows only) - there's no longer a need to mess with VcXsrv config files to get the right keyboard. Fixes #1729, #2065, and probably #2156.
- Fixes #2096 (preview windows when hovering over taskbar in Windows)
- Window titles (on Windows) are displayed properly - fixes #2590
- RGBA cursors can now be used (if someone wants to make them)
- The palette windows can now be docked/undocked within a single session on OS X (due to removal of reparenting code)
- Technically supports running natively on Wayland by virtue of GDK supporting it - but this is untested - and server side decorations would have to be provided, which I don't think happens as of now.
- Apparently better performance under OS X, especially when scrolling (XQuartz performance related issues)

Technically the GUI can now also be built on Homebrew using this backend (#2668), although I don't have enough experience with the OS X packaging. 
#### Caveats

Expect this backend to crash a bit more than on X11. I have been testing quite a lot, but it's difficult to emulate the behaviour of that backend exactly. There may be differences in behaviour that I may have overlooked which causes odd things to happen.
### Types of changes

Changes to FontForge outside of adding this backend have been kept as minimal as possible. However, some changes did have to be made:
- Removal of XOR mode drawing. It's simply not supported by Cairo, which is the only way to draw things when using GDK. Luckily this only affects a few things - the cursor blinking, the ruler position markers and the guidelines in the charview. I've made changes to all three of these to make sure that they still work (either by doing expose calls - or in the case of the cursor, using 'difference' mode). There may be more efficient ways to do this.
  - Update 2016/09/12: I've made rulers get drawn to an offscreen pixmap which then gets drawn to the screen. When the marker position changes, I copy part of that pixmap back over the screen where the old marker was before drawing the new marker. Performance is now roughly equivalent to what it was with just XOR mode drawing.
- Removal of all code that used the reparenting feature - which was only the palette toolbars. GDK's support of this was flakey - especially on OS X. Even XQuartz couldn't do it right (try undocking then docking the palettes on a current version of FF). Anyway, this was fixed by simply destroying then re-creating the palettes whenever it's docked/undocked, which I don't think is a big deal.
- I implemented the `DrawArc` function in Cairo for the X11 backend.
- Popups (tooltips) no longer show if FF doesn't have focus 
- Undocked palette windows now behave properly especially on Windows - they remain above the charview even if the charview has focus.
### Known remaining issues
- I still don't understand the start routine for OS X. To be honest, `startui.c` is in particularly bad shape. I remember #1277. What I do know is that it has a separate event loop to Windows and Linux, which seems odd.
- I don't know if resolution is detected correctly on OS X (particularly on HiDPI displays). GDK has a function ([gdk_screen_get_resolution](https://developer.gnome.org/gdk3/stable/GdkScreen.html#gdk-screen-get-resolution)) to get it, but on my test rig (_cough_) it reports 72dpi, even though it should be rendered at 96dpi. Anyway, I added a property called "ScreenResolution", so this value can be manually set (no messing around with "ScreenWidth..." settings).
- The clipboard should work on all three platforms. Note however, that copy/pasting to/from external applications on OS X is limited to text only.
- Drag and drop to/from external applications is not supported. Although I'm led to believe that this didn't actually work with the current version anyway. 
- IME support doesn't exist on OS X/Linux. Unlikely to change because this support lies in the GTK layer. 
- The palette windows now lack a black border around them when docked - they relied on the window manager to do this (as reparented toplevel windows), but this no longer happens because they're implemented as child windows now.
- Minimum size windows (GDK constraint vs FF asking for ridiculous min widths)
- Modal focus is sometimes broken - FF allows for some really weird window combinations.
- Repeated mouse events when the mouse isn't moving is a bug(difference?) of VcXsrv on Windows - can reproduce with FF running on Linux X-forwarded to Windows

### CODE REVIEW

**Comment by mskala** [comment]
I'd be happier to see a move away from GDraw than a patch to make it easier to continue using GDraw, but this is impressive work and if it makes the package play better with desktop environments, it's a good thing.

**Comment by jtanx** [comment]
IMO, the only option of moving away from GDraw would be to move to Qt. Now I've had time to see it, I hate GTK (in particular GTK3). In any case, that would be a major undertaking, far greater than this patch would ever be.

**Comment by jtanx** [comment]
Here's a GDK2 build for Windows: https://ci.appveyor.com/project/fontforge/fontforge/build/1.0.111/artifacts (I set it to GDK2 because I normally build GDK3 and I want to ensure it still builds on GDK2)

Some noted differences between GDK2 and GDK3:
* GDK2 and GDK3 seem to have different drawing models. On GDK3, invalidating a child window (which can happen by calling gdk_window_invalidate_region, or by resizing or raising the window) will cause the parent window to also be invalidated, which results in flicker. The same doesn't happen on GDK2
* GDK3 appears to have better drawing performance, at least on Windows
* The API instability on GDK3 is insane. Functions deprecated left right and centre. Not looking forward to GDK 3.22, although apparently that's meant to be the one with a stable API.
*GDK3 can apply motion compression whereas the same cannot be done easily on GDK2 (this was the point of `GDrawSkipMouseEvents`, but GDK doesn't really provide much in the way of queue manipulation). This means that on complex glyphs, using GDK2 can appear to lag behind.

Known issues remaining:
* Ctrl clicking + dragging in the charview lags, especially on GDK2. Yet to investigate, but the hand tool does some pretty stupid things like invalidate the whole window if the rulers are enabled.
* Sometimes modals break things - such as right clicking in font view -> expand stroke. The way FontForge handles modals is... interesting.

**Comment by jtanx** [comment]
Uh. So I noticed a slow-down when you zoomed in too far on a glyph - see [here](https://github.com/fontforge/fontforge/projects/1#card-98098). I had fixed a bug where a path that was indicated as to be drawn with a dashed line was not actually drawn as such (see [here](https://github.com/fontforge/fontforge/pull/2782/files#diff-9148cd615a33f3feaca272dc1bb6f7a2R176). 

So old FF was drawing a solid line while I was drawing a dashed line. It turns out if you zoom in too far on a dashed line, Cairo runs slow.

stroke

So.. does anyone know what the dashed vs solid line is meant to mean?

**Comment by pathumego** [comment]
@jtanx I can arrange testing on Win and Mac on a daily workflow if you require.

**Comment by davelab6** [comment]
@jtanx this is awesome work!!

**Comment by jtanx** [comment]
gnulib is broken again.

**Comment by davelab6** [comment]
@jtanx wrote, 

> So... does anyone know what the dashed vs solid line is meant to mean?

That's odd. I thought it was meant to be solid, not dashed.

**Comment by davelab6** [comment]
@pathumego did you check out https://ci.appveyor.com/project/fontforge/fontforge/build/1.0.111/artifacts ? :)

File: gdraw/ggdkcdraw.c [hunk 1]
```diff
@@ -51,7 +51,16 @@ static void _GGDKDraw_CheckAutoPaint(GGDKWindow gw) {
             _GGDKDraw_CleanupAutoPaint(gw->display);
             gw->display->dirty_window = gw;
         }
+
+        // Shut up! I know what I'm doing.
+#ifdef GGDKDRAW_GDK_3_22
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+#endif
```
Comment (khaledhosny) [inline gdraw/ggdkcdraw.c]: You can simply use `G_GNUC_BEGIN_IGNORE_DEPRECATIONS`/`G_GNUC_END_IGNORE_DEPRECATIONS` from glib.

Comment (jtanx) [reply gdraw/ggdkcdraw.c]: Thanks.

**Comment by khaledhosny** [comment]
> I don't know if resolution is detected correctly on OS X (particularly on HiDPI displays). GDK has a function (gdk_screen_get_resolution) to get it, but on my test rig (cough) it reports 72dpi, even though it should be rendered at 96dpi. Anyway, I added a property called "ScreenResolution", so this value can be manually set (no messing around with "ScreenWidth..." settings).

Is this for hi-dpi auto detection? You may want to use `gdk_monitor_get_scale_factor ()` for this instead.

**Comment by khaledhosny** [comment]
I just tried this on Linux and it is working on hi-dpi screen without any manual settings:
image
(The glyphs in the fontview are blurry, but that is probably they did not take the scale factor into account when rasterized, and Cairo just scaled the bitmaps.)

**Comment by khaledhosny** [comment]
Any updates here? Since this keeps the default X backend essentially unchanged, I think it should be merged and further work on GDK backend can be continued (potentially allowing others to contribute).

**Comment by jtanx** [comment]
Sorry about the delay, yeah I'm fairly happy to merge it (any objections, @frank-trampe?); it more or less works. I got busier, and then I saw the breaking changes coming in GDK4. That coupled with debating what version of GDK to support, or if this would have been possible/better done at the GTK level, and I kind of got put off by making any further changes.

I'll see if I can clean it up a little before merging though.

**Comment by khaledhosny** [comment]
I think GTK3/GDK3 will be around for some time and the API is now stabilized, so that is a win. GTK3 was first released in 2011 and GTK2 is still a thing today.

Doing this work with GTK instead of GDK is something to explore indeed, but I think it would be easier to do on top of this work.

**Comment by davelab6** [comment]
Agree with Khaled on all points

On Mar 10, 2018 10:10 AM, "Khaled Hosny" <notifications@github.com> wrote:

I think GTK3/GDK3 will be around for some time and the API is now
stabilized, so that is a win. GTK3 was first released in 2011 and GTK2 is
still a thing today.

Doing this work with GTK instead of GDK is something to explore indeed, but
I think it would be easier to do on top of this work.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/fontforge/fontforge/pull/2782#issuecomment-372015557>,
or mute the thread

**Comment by jtanx** [comment]
Thanks for testing! Yeah sure, feel free to create issues for anything you find.

**Comment by MaddTheSane** [comment]
> GDK has a function (gdk_screen_get_resolution) to get [the desktop DPI], but on my test rig (cough) it reports 72dpi, even though it should be rendered at 96dpi.

For historical reasons, Mac OS X uses 72 DPI for the base 1x resolution.

**Pull request merged**